src/dpapi_ng/_client.py

```diff
@@ -226,6 +230,65 @@ def _decrypt_blob(
     )
 
 
+def _encrypt_blob(
+    blob: bytes,
+    key: GroupKeyEnvelope,
+    security_descriptor: bytes,
+    sid: str,
+) -> bytes:
+    # Generate cek and encrypt our payload.
+    cek = AESGCM.generate_key(bit_length=256)
+    cek_iv = os.urandom(12)
+
+    parameters_writer = ASN1Writer()
+    with parameters_writer.push_sequence() as parameters:
+        parameters.write_octet_string(cek_iv)
+        parameters.write_integer(16)
+
+    enc_content_algorithm = "2.16.840.1.101.3.4.1.46"
+    enc_content_parameters = parameters_writer.get_data()
+    enc_content = content_encrypt(
+        enc_content_algorithm,
+        enc_content_parameters,
+        cek,
+        blob,
+    )
+
+    # Wrap the cek with the kek.
+    key_identifier = KeyIdentifier(
+        version=1,
+        flags=key.flags,
+        l0=key.l0,
+        l1=key.l1,
+        l2=key.l2,
+        root_key_identifier=key.root_key_identifier,
+        key_info=b'',
```

Comment:
This will need to be the public key `key.l2_key` if `key.is_public_key` otherwise it's a special nonce used for the CEK decryption on the client side. I'm surprised Windows would be able to decrypt the value if this was not set.

Reply:
Turns out that this is only working if the authenticated user is member of the target group (SID). If not, ["Current user is not authorized to retrieve the KEK information"](https://github.com/jborean93/dpapi-ng/blob/main/src/dpapi_ng/_gkdi.py#L499) is thrown. When commenting out this `raise ValueError` and using `key.l2_key` as `key_info`, the script encrypts without any errors but decryption does not work: `cryptography.hazmat.primitives.keywrap.InvalidUnwrap in _crypto.py, line 34`. Do you have any idea how to fix this?

Reply:
If the user isn't authorised to get the actual seed key from `GetKey` it will return the public key used with the L group identifier. This public key is what is embedded in the `key_info` rather than the special nonce that is used if it's not a public key.

Reply:
Yes, I tried to use this public key as `key_info` but trying to decrypt the resulting encrypted blob does not work. I'm currently unable to find the problem.

Reply:
Ah I see the conundrum. The current code calls `GroupKeyEnvelope.get_key` which only works if it contains the L seed key rather than the public key. I'll have to look into the code to see how Microsoft still produces a valid KEK here as the current code relies on knowing the private key.

Reply:
I got it working for DH keys. Essentially you need to generate your own private and public key pair and include that in the key identifier. The public key returned by GetKey is the peers public key which gives us enough information to generate the KEK. The following diff on top of your latest commit adds support for this scenario with the DH secret algorithm. I haven't looked into the ECDH_P* algorithms yet but the logic should be roughly the same. I also tweaked the nonce to actually be a random 32-byte value rather than the empty byte string to match what Windows does.

<details>
  <summary>Click me to expand diff</summary>

```diff
diff --git a/src/dpapi_ng/_client.py b/src/dpapi_ng/_client.py
index 4a09d70..a6551f9 100644
--- a/src/dpapi_ng/_client.py
+++ b/src/dpapi_ng/_client.py
@@ -6,10 +6,8 @@ from __future__ import annotations
 import typing as t
 import uuid
 
-from cryptography.hazmat.primitives.ciphers.aead import AESGCM
-
 from ._asn1 import ASN1Writer
-from ._blob import DPAPINGBlob, KeyIdentifier
+from ._blob import DPAPINGBlob
 from ._crypto import (
     AlgorithmOID,
     cek_decrypt,
@@ -261,20 +259,7 @@ def _encrypt_blob(
         blob,
     )
 
-    # Wrap the cek with the kek.
-    key_identifier = KeyIdentifier(
-        version=1,
-        flags=key.flags,
-        l0=key.l0,
-        l1=key.l1,
-        l2=key.l2,
-        root_key_identifier=key.root_key_identifier,
-        key_info=b"",
-        domain_name=key.domain_name,
-        forest_name=key.forest_name,
-    )
-    kek = key.get_kek(key_identifier)
-
+    kek, key_identifier = key.new_kek()
     enc_cek_parameters = None
     enc_cek = cek_encrypt(
         enc_cek_algorithm,
diff --git a/src/dpapi_ng/_gkdi.py b/src/dpapi_ng/_gkdi.py
index 92f6341..6f4183d 100644
--- a/src/dpapi_ng/_gkdi.py
+++ b/src/dpapi_ng/_gkdi.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 
 import dataclasses
 import math
+import os
 import typing as t
 import uuid
 
@@ -526,6 +527,55 @@ class GroupKeyEnvelope:
                 32,
             )
 
+    def new_kek(
+        self,
+    ) -> tuple[bytes, KeyIdentifier]:
+        if self.kdf_algorithm != "SP800_108_CTR_HMAC":
+            raise NotImplementedError(f"Unknown KDF algorithm '{self.kdf_algorithm}'")
+
+        kdf_parameters = KDFParameters.unpack(self.kdf_parameters)
+        hash_algo = kdf_parameters.hash_algorithm
+
+        if self.is_public_key:
+            # If is_public_key flag is set, the L2 key is the peer's public key
+            private_key = os.urandom(math.ceil(self.private_key_length / 8))
+            kek = compute_kek(
+                algorithm=hash_algo,
+                secret_algorithm=self.secret_algorithm,
+                secret_parameters=self.secret_parameters,
+                private_key=private_key,
+                public_key=self.l2_key,
+            )
+
+            key_info = compute_public_key(
+                secret_algorithm=self.secret_algorithm,
+                secret_parameters=self.secret_parameters,
+                private_key=private_key,
+                peer_public_key=self.l2_key,
+            )
+        else:
+            key_info = os.urandom(32)
+            kek = kdf(
+                hash_algo,
+                self.l2_key,
+                KDS_SERVICE_LABEL,
+                key_info,
+                32,
+            )
+
+        key_identifier = KeyIdentifier(
+            version=1,
+            flags=self.flags,
+            l0=self.l0,
+            l1=self.l1,
+            l2=self.l2,
+            root_key_identifier=self.root_key_identifier,
+            key_info=key_info,
+            domain_name=self.domain_name,
+            forest_name=self.forest_name,
+        )
+        return kek, key_identifier
+
     @classmethod
     def unpack(
         cls,
@@ -733,10 +783,6 @@ def compute_kek_from_public_key(
     public_key: bytes,
     private_key_length: int,
 ) -> bytes:
-    # Special thanks for Grzegorz Tworek (@0gtweet) and Michał Grzegorzewski
-    # for providing access to CQDPAPINGPFXDecrypter.exe which contains the
-    # BCrypt* APIs Microsoft use to derive the KEK.
-
     private_key = kdf(
         algorithm,
         seed,
@@ -745,6 +791,26 @@ def compute_kek_from_public_key(
         private_key_length,
     )
 
+    return compute_kek(
+        algorithm,
+        secret_algorithm=secret_algorithm,
+        secret_parameters=secret_parameters,
+        private_key=private_key,
+        public_key=public_key,
+    )
+
+
+def compute_kek(
+    algorithm: hashes.HashAlgorithm,
+    secret_algorithm: str,
+    secret_parameters: t.Optional[bytes],
+    private_key: bytes,
+    public_key: bytes,
+) -> bytes:
+    # Special thanks for Grzegorz Tworek (@0gtweet) and Michał Grzegorzewski
+    # for providing access to CQDPAPINGPFXDecrypter.exe which contains the
+    # BCrypt* APIs Microsoft use to derive the KEK.
+
     secret_hash_algorithm: hashes.HashAlgorithm
     if secret_algorithm == "DH":
         # p = FFCDHParameters.unpack(secret_parameters or b"")
@@ -798,3 +864,33 @@ def compute_kek_from_public_key(
         kek_context,
         32,
     )
+
+
+def compute_public_key(
+    secret_algorithm: str,
+    secret_parameters: t.Optional[bytes],
+    private_key: bytes,
+    peer_public_key: bytes,
+) -> bytes:
+    if secret_algorithm == "DH":
+        dh_pub_key = FFCDHKey.unpack(peer_public_key)
+
+        # We can derive our public key based on the DH formula.
+        # X = G**x mod p
+        my_pub_key = pow(
+            dh_pub_key.generator,
+            int.from_bytes(private_key, byteorder="big"),
+            dh_pub_key.field_order,
+        )
+        return FFCDHKey(
+            dh_pub_key.key_length,
+            dh_pub_key.field_order,
+            dh_pub_key.generator,
+            my_pub_key,
+        ).pack()
+
+    elif secret_algorithm.startswith("ECDH_P"):
+        raise NotImplementedError("FIXME implement support for ECDH keys")
+
+    else:
+        raise NotImplementedError(f"Unknown secret agreement algorithm '{secret_algorithm}'")
```
</details>

Reply:
Very nice, thank you!